pkg/cluster/loadbalancerprofile.go

```diff
@@ -263,8 +317,31 @@ func (m *manager) reconcileDesiredManagedIPs(ctx context.Context) ([]api.Resourc
 			break
 		}
 	}
+	return desiredIPAddresses
+}
 
-	return desiredIPAddresses, nil
+func (m *manager) createPublicIPAddresses(ctx context.Context, ipAddresses map[string]mgmtnetwork.PublicIPAddress, numToCreate int) error {
+	ch := make(chan createIPResult)
+	defer close(ch)
+	var errResults []string
+	// create additional IPs if needed
+	for i := 0; i < numToCreate; i++ {
+		go m.createPublicIPAddress(ctx, ch)
```

Comment:
More precisely, _pkg/cluster/loadbalancerprofile_test.go:114_ is what is modifying the global variable and the fake generator is not concurrent safe. Even this test line was not introduced in this PR, we are now using it concurrently.

The quick fix is to make fake generator concurrent safe by using a mutex:

```go
package fake

// Copyright (c) Microsoft Corporation.
// Licensed under the Apache License 2.0.

import (
	"sync"

	"github.com/Azure/ARO-RP/pkg/util/uuid"
)

type fakeGenerator struct {
	words      []string
	currentPos int
	mu         *sync.Mutex
}

func NewGenerator(predefinedWords []string) uuid.Generator {
	return &fakeGenerator{
		words: predefinedWords,
		mu:    &sync.Mutex{},
	}
}

func (f *fakeGenerator) movePos() {
	f.currentPos++
}

func (f *fakeGenerator) Generate() string {
	f.mu.Lock()
	defer f.mu.Unlock()

	defer f.movePos()
	if len(f.words) < f.currentPos {
		return ""
	}
	return f.words[f.currentPos]
}

```

Reply:
>I checked locally the PR and executed
> `go test -race ./pkg/cluster` 

Should we be doing this as part of our normal test suite?

Reply:
> Should we be doing this as part of our normal test suite? 
That is my question too, I was unaware of this check so it may be best to add this to CI.

Reply:
Good question 🙂 
I mentioned in the past that we have some race conditions in some test helper functions and utilities. Adding the -race flag is something that should be done when working with concurrent code, but we can not include it as a default step in the CI as it will fail with those "expected" race conditions we have in the repo. We would ideally remove the race conditions and later make that -race option a default. 
This is the [PR](https://github.com/Azure/ARO-RP/pull/2466) where I fixed some, but I was not able to fix all of them in a reasonable time.